### PR TITLE
Fix: Added sharding Constraints for MTP block (b/481469708)

### PR DESCRIPTION
**Pull request by parambole**
# Description

The fix involves adding a single, targeted sharding annotation to `target_token_embedding` within the `MultiTokenPredictionLayer.__call__` method using `sharding.maybe_shard_with_logical`. This ensures that sharding is correctly propagated through the MTP layer's internal operations, including the projection and the subsequent transformer layer.

FIXES: b/481469708

# Tests

Verified with `train_compile.py` on the TPU backend and validated on ironwood run, where the model now trains without erros.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by NuojCheng — APPROVED** [review]

File: src/MaxText/layers/multi_token_prediction.py [hunk 1]
```diff
@@ -118,10 +119,11 @@ def __init__(
     self.transformer_layer = nnx_wrappers.ToNNX(mtp_transformer_layer, rngs=rngs)
 
     # ToNNX requires explicit initialization with sample inputs for proper parameter setup.
+    batch_size, seq_len = max_utils.get_batch_seq_len_for_mode(config=cfg, model_mode=MODEL_MODE_TRAIN)
     self.transformer_layer.lazy_init(
-        inputs=jnp.zeros((1, 1, cfg.base_emb_dim), dtype=cfg.dtype),
+        inputs=jnp.zeros((batch_size, seq_len, self.config.base_emb_dim), dtype=self.config.dtype),
```
Comment (RissyRan) [inline src/MaxText/layers/multi_token_prediction.py]: Usually we use `self.config.emb_dim` instead of base_emb_dim. Potentially leverage the global scaling: https://github.com/AI-Hypercomputer/maxtext/blob/55b57ffe957c9a36c42f4ac730ca35849115e9ab/src/maxtext/configs/base.yml#L144

Comment (parambole) [reply src/MaxText/layers/multi_token_prediction.py]: Thank you for the suggestion. Done!
